machinist/mserver/controller.go

```diff
@@ -7,11 +7,22 @@ import (
 )
 
 type Controller struct {
-	workers        sync.Map
-	ConnectedNodes map[string][]string
+	connectedNodes map[string]*Node
 	sync.RWMutex
 }
 
+func (en *Controller) Nodes() []*Node {
```

Comment:
nit, but important: I generally assume that public methods of a function are ~always safe to call, or prominently indicate they aren't.

Nodes() here can only be called by someone holding the lock. Can you mark the method private AND add a comment indicating that the caller is expected to hold the Controller Lock?

Reply:
Or make it public, with proper locking.